### PR TITLE
Merge new buffer implementation API

### PR DESCRIPTION
**Pull request by kg583**
Components of `TIVar` objects will now be `Buffers`, which implement the descriptor protocol to permit users to interact with those components as if they were their canonical types, while also protecting against improper data. The new implementation also streamlines the reading and writing API and utilizes a sleeker subclassing setup.

### CODE REVIEW

**Comment by kg583** [comment]
There has been a suggestion to replace `load_` with `from_` and prepend the matching methods with `to_`. Any particular feelings one way or the other about this are welcome.

File: tivars/models/__init__.py [hunk 1]
```diff
@@ -2,7 +2,7 @@ class TIModel:
     def __init__(self, name: str, flags: int, signature: str):
```
Comment (Legend-of-iPhoenix) [inline tivars/models/__init__.py:2]: I'm concerned that this signature field here might be conflated with a cryptographic signature or something else; I think it's more appropriate to break away from what tivarslibcpp does here and call it `file_magic` or just `magic`, see https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files

**Review by Legend-of-iPhoenix — CHANGES_REQUESTED** [review]

File: tivars/models/__init__.py (hunk 1 above)
Comment (kg583) [reply tivars/models/__init__.py:2]: "Magic" feels rather non-descriptive, even owing to the term's precedence, but I guess it isn't potentially actively confusing

Comment (Legend-of-iPhoenix) [reply tivars/models/__init__.py:2]: yeah, it's not the greatest, but it is an arbitrary number ("magic number") that has meaning

Comment (Legend-of-iPhoenix) [reply tivars/models/__init__.py:2]: delete this

File: tivars/var.py [hunk 2]
```diff
@@ -111,137 +61,163 @@ def checksum(self) -> int:
...
-        self.comment = file.read(42).decode('utf8')
-        entry_length = int.from_bytes(file.read(2), 'little')
+        self.load_buffer(TIVar.signature, data)
+        match self.signature:
+            case TI_82.signature:
+                model = TI_82
+            case TI_83.signature:
+                model = TI_83
+            case TI_84p.signature:
+                model = TI_84p
```
Comment (Legend-of-iPhoenix) [inline tivars/var.py:159]: Wouldn't it be less failure-prone to assume the model with the most feature bits; i.e. the python CE?

File: tivars/var.py [hunk 3]
```diff
@@ -111,137 +61,163 @@ def checksum(self) -> int:
...
+        elif type_id != self.type_id:
+            raise TypeError("The var type is incorrect. Use a TIVar instance if you don't know the type.")
+
+        self.load_buffer(TIVar.name, data)
 
         if self.meta_length == TIVar.flash_meta_length:
-            self.version = file.read(1)
-            self.archived = file.read(1) != b'\x00'
+            if self.model is not None and not self.model.has(TIFeature.FLASH):
+                warnings.warn(f"Calculator model does ")
```
Comment (Legend-of-iPhoenix) [inline tivars/var.py:189]: missing "not support flash"

perhaps even use the .archive() method here

File: tivars/var.py [hunk 4]
```diff
@@ -111,137 +61,163 @@ def checksum(self) -> int:
...
+            self.load_buffer(TIVar.version, data)
+            self.load_buffer(TIVar.archived, data)
 
         elif self.meta_length != TIVar.base_meta_length:
             warnings.warn(f"The var entry meta length has an unexpected value ({self.meta_length}).",
                           BytesWarning)
-            self.corrupt = True
 
-        data_length2 = int.from_bytes(file.read(2), 'little')
+        data_length2 = int.from_bytes([next(data), next(data)], 'little')
```
Comment (Legend-of-iPhoenix) [inline tivars/var.py:198]: wasn't this one of the things we hated about the cpp implementation

File: tivars/var.py (hunk 2 above)
Comment (kg583) [reply tivars/var.py:159]: Probably

File: tivars/var.py (hunk 3 above)
Comment (kg583) [reply tivars/var.py:189]: Ope; also I need to set the version bits conditionally anyway so

File: tivars/var.py (hunk 4 above)
Comment (kg583) [reply tivars/var.py:198]: `data_length2` isn't actually stored anywhere; its just used for file verification.

File: tivars/models/__init__.py (hunk 1 above)
Comment (kg583) [reply tivars/models/__init__.py:2]: Yeah, I'm on board with this, its just a matter of deciding between `magic` or `file_magic` or some secret third thing

**Review by Legend-of-iPhoenix — APPROVED** [review]

**Comment by kg583** [comment]
Do not merge until more tests have been written.

**Comment by kg583** [comment]
> Are we using "TI-" for the model name strings or not?
> 
> it's either a commit here or a commit to https://github.com/CE-Fauxgramming/tokens

We are.

**Pull request merged**